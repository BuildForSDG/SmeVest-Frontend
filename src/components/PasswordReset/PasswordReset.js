import React from 'react';
import { connect } from 'react-redux';
import { Link, withRouter } from 'react-router-dom';
import {
  Container, Row, Col, Form, Alert, Spinner,
} from 'react-bootstrap';
import {
  checkServerNetworkError,
  checkServerEmailError,
  checkError,
  checkValidity,
  isFormValidCheck,
} from '../../utils/validation';
import { clearAuthErrors, resetPassword } from '../../redux/actions';

import './PasswordReset.css';

export class PasswordReset extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      email: {
        value: '',
        valid: false,
        validation: {
          required: true,
          email: true,
        },
        touched: false,
      },
      password: {
        value: '',
        valid: false,
        validation: {
          required: true,
          minLength: 6,
        },
        touched: false,
      },
      confirm: {
        value: '',
        valid: false,
        validation: {
          required: true,
          match: 'password',
        },
        touched: false,
      },
      token: '',
      isFormValid: false,
    };

    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  componentDidMount() {
    // Get token from url
    const { token } = this.props.match.params;
    if (!token.trim()) {
      this.props.history.push('/');
    } else {
      this.setState((prevState) => ({ ...prevState, token }));
    }
  }

  componentWillUnmount() {
    this.props.clearStateErrors();
  }

  handleInputChange(e) {
    e.preventDefault();
    const stateToUpdate = { ...this.state };
    const inputElement = stateToUpdate[e.target.name];
    inputElement.value = e.target.value;
    inputElement.valid = checkValidity(e.target.value, inputElement.validation, this.state);
    inputElement.touched = true;
    stateToUpdate[e.target.name] = inputElement;

    const isFormValid = isFormValidCheck(stateToUpdate);

    this.setState({ ...stateToUpdate, isFormValid });
  }

  async handleSubmit(e) {
    e.preventDefault();
    const { email, password, token } = this.state;
    const user = { email: email.value, password: password.value, token };
    const res = await this.props.reset(user);
    if (res) {
      this.setState({
        success: true,
        email: '',
        password: '',
        confirm: '',
      });
    }
  }

  render() {
    return (
      <Container className="PasswordReset">
        <Row>
          <Col md={{ span: 6, offset: 3 }}>
            <Form className="Form">
              {this.state.success ? (
                <Alert variant="success">
                  <Alert.Heading>Password reset is successful. Proceed to sign in.</Alert.Heading>
                  <Link to="/signin">Sign In</Link>
                </Alert>
              ) : checkServerNetworkError(this.props.error) ? (
                <Alert variant="danger">{this.props.error.network}</Alert>
              ) : (
                <Alert variant="info">Enter your email and new password</Alert>
              )}
              <Form.Group controlId="formGroupEmail">
                <Form.Label>Email address</Form.Label>
                <Form.Control
                  type="email"
                  name="email"
                  value={this.state.email.value}
                  onChange={this.handleInputChange}
                  placeholder="Enter email"
                  className={`${checkServerEmailError(this.props.error) ? 'is-invalid' : checkError(this.state.email)}`}
                />
                <Form.Control.Feedback type={`${this.state.email.valid && !this.props.error ? 'valid' : 'invalid'}`}>
                  {`${
                    checkServerEmailError(this.props.error)
                      ? this.props.error.email
                      : this.state.email.valid
                        ? 'Email looks good'
                        : 'Invalid email address'
                  }`}
                </Form.Control.Feedback>
              </Form.Group>
              <Form.Group controlId="formGroupPassword">
                <Form.Label>Password</Form.Label>
                <Form.Control
                  type="password"
                  name="password"
                  value={this.state.password.value}
                  onChange={this.handleInputChange}
                  placeholder="Password"
                  className={`${checkError(this.state.password)}`}
                />
                <Form.Control.Feedback type={`${this.state.password.valid ? 'valid' : 'invalid'}`}>
                  {`${this.state.password.valid ? 'Passowrd looks good' : 'Password min length is 6'}`}
                </Form.Control.Feedback>
              </Form.Group>
              <Form.Group controlId="formGroupConfirmPassword">
                <Form.Label>Confirm Password</Form.Label>
                <Form.Control
                  type="password"
                  name="confirm"
                  value={this.state.confirm.value}
                  onChange={this.handleInputChange}
                  placeholder="Confirm Password"
                  className={`${checkError(this.state.confirm)}`}
                />
                <Form.Control.Feedback type={`${this.state.confirm.valid ? 'valid' : 'invalid'}`}>
                  {`${this.state.confirm.valid ? 'Password match' : 'Password does not match'}`}
                </Form.Control.Feedback>
              </Form.Group>
              <Form.Group>
                {this.props.loading && <Spinner animation="border" className="float-right" variant="info" />}
                {!this.props.loading && (
                  <button
                    disabled={!this.state.isFormValid}
                    onClick={this.handleSubmit}
                    type="submit"
                    className="Auth-Button float-right"
                  >
                    Reset
                  </button>
                )}
              </Form.Group>
            </Form>
          </Col>
        </Row>
      </Container>
    );
  }
}

const mapStateToProps = (state) => ({
  loading: state.auth.loading,
  error: state.auth.error,
});

const mapDispatchToProps = (dispatch) => ({
  clearStateErrors: () => dispatch(clearAuthErrors()),
  reset: ({ email, password, token }) => dispatch(resetPassword({ email, password, token })),
});

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(PasswordReset));
