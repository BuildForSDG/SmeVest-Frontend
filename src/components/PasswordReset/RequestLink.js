import React from 'react';
import { connect } from 'react-redux';
import { Link, withRouter } from 'react-router-dom';
import {
  Container, Row, Col, Form, Alert, Spinner,
} from 'react-bootstrap';
import { checkServerNetworkError } from '../../utils/validation';
import { clearAuthErrors, requestResetLink } from '../../redux/actions';
import './PasswordReset.css';

export class RequestLink extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      success: false,
    };

    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  componentWillUnmount() {
    this.props.clearStateErrors();
  }

  handleInputChange(e) {
    e.preventDefault();
    this.setState({ email: e.target.value });
  }

  async handleSubmit(e) {
    e.preventDefault();
    const res = await this.props.requestLink(this.state.email);
    if (res) {
      this.setState({ success: true, email: '' });
    }
  }

  render() {
    return (
      <Container className="PasswordReset">
        <Row>
          <Col md={{ span: 6, offset: 3 }}>
            <Form className="Form">
              {this.state.success ? (
                <Alert variant="success">
                  <Alert.Heading>
                    A link has been sent to your email address.{'\n'}
                    Follow the link to reset your password.
                  </Alert.Heading>
                </Alert>
              ) : checkServerNetworkError(this.props.error) ? (
                <Alert variant="danger">{this.props.error.network}</Alert>
              ) : (
                <Alert variant="info">
                  <Alert.Heading> Forgot your password?</Alert.Heading>
                  Please enter your email address below. We&apos;ll{'\n'}
                  send you an email with a link to reset your password. <Link to="/">Cancel</Link>
                </Alert>
              )}
              <Form.Group controlId="formGroupEmail">
                <Form.Label>Email Address</Form.Label>
                <Form.Control
                  type="email"
                  name="email"
                  value={this.state.email}
                  onChange={this.handleInputChange}
                  placeholder="Enter email address"
                />
              </Form.Group>
              <Form.Group>
                {this.props.loading && <Spinner animation="border" className="float-right" variant="info" />}
                {!this.props.loading && (
                  <button
                    disabled={!this.state.email}
                    onClick={this.handleSubmit}
                    type="submit"
                    className="Auth-Button float-right"
                  >
                    Send
                  </button>
                )}
              </Form.Group>
            </Form>
          </Col>
        </Row>
      </Container>
    );
  }
}

const mapStateToProps = (state) => ({
  loading: state.auth.loading,
  error: state.auth.error,
});

const mapDispatchToProps = (dispatch) => ({
  clearStateErrors: () => dispatch(clearAuthErrors()),
  requestLink: (email) => dispatch(requestResetLink(email)),
});

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(RequestLink));
