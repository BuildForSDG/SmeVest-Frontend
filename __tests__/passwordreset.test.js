import React from 'react';
import { shallow } from 'enzyme';
import { PasswordReset } from '../src/components/PasswordReset/PasswordReset';

describe('PasswordReset Component', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<PasswordReset />);
  });

  test('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
