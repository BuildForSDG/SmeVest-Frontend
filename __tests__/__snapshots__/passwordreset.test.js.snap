// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordReset Component should match snapshot 1`] = `ShallowWrapper {}`;
